Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116688
* __->__ #116413

**Summary**
As per the discussion with Jerry in https://github.com/pytorch/pytorch/pull/114540#discussion_r1406945902, we plan to refactor `X86InductorQuantizer` annotations with `SubgraphMatcherWithNameNodeMap`. This plan will be implemented through several PRs, with this one being the first, focusing on refactoring single conv2d annotations in PTQ.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

leslie-fang-intel committed:
Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap"


**Summary**
As per the discussion with Jerry in https://github.com/pytorch/pytorch/pull/114540#discussion_r1406945902, we plan to refactor `X86InductorQuantizer` annotations with `SubgraphMatcherWithNameNodeMap`. This plan will be implemented through several PRs, with this one being the first, focusing on refactoring single conv2d annotations in PTQ.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap"


**Summary**
As per the discussion with Jerry in https://github.com/pytorch/pytorch/pull/114540#discussion_r1406945902, we plan to refactor `X86InductorQuantizer` annotations with `SubgraphMatcherWithNameNodeMap`. This plan will be implemented through several PRs, with this one being the first, focusing on refactoring single conv2d annotations in PTQ.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jerryzh168 commented:
great! maybe you can test different conv configurations to make sure the pattern can catch all of them, including different kernel_size/stride/padding and symbolic shapes

leslie-fang-intel committed:
Update on "Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap"


**Summary**
As per the discussion with Jerry in https://github.com/pytorch/pytorch/pull/114540#discussion_r1406945902, we plan to refactor `X86InductorQuantizer` annotations with `SubgraphMatcherWithNameNodeMap`. This plan will be implemented through several PRs, with this one being the first, focusing on refactoring single conv2d annotations in PTQ.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jgong5 commented on torch/ao/quantization/quantizer/utils.py:
```diff
@@ -47,3 +60,55 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
+    # https://github.com/pytorch/pytorch/blob/
+    # 4c6e842496da636123f83ef868ca1974631f1f1e/torch/_dynamo/config.py#L35-L39
+    return _generate_pattern_matcher_helper(
+        _get_conv_unary_pattern(torch.nn.functional.conv2d),
```
How is `torch.nn.Conv2d` handled?

leslie-fang-intel replied:
I think `torch.nn.Conv2d` and `torch.nn.functional.conv2d` will generate same node `torch.ops.aten.conv2d.default` after export. So, we should be fine to only use `torch.nn.functional.conv2d` here.

leslie-fang-intel committed:
Update on "Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap"


**Summary**
As per the discussion with Jerry in https://github.com/pytorch/pytorch/pull/114540#discussion_r1406945902, we plan to refactor `X86InductorQuantizer` annotations with `SubgraphMatcherWithNameNodeMap`. This plan will be implemented through several PRs, with this one being the first, focusing on refactoring single conv2d annotations in PTQ.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

leslie-fang-intel commented:
> great! maybe you can test different conv configurations to make sure the pattern can catch all of them, including different kernel_size/stride/padding and symbolic shapes

- Add the testing for different `kernel_size/stride/padding`
- For `symbolic shapes`, did you mean add `constraints` parameters in `capture_pre_autograd_graph`? I think there should be no `symbolic size` for the pattern generated by Conv2D.

jerryzh168 commented:
> * For `symbolic shapes`, did you mean add `constraints` parameters in `capture_pre_autograd_graph`? I think there should be no `symbolic size` for the pattern generated by Conv2D.

Yeah, but we may still want to test it to make sure it also works

jerryzh168 commented on torch/ao/quantization/quantizer/utils.py:
```diff
@@ -47,3 +60,55 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
     )
+
+
+def _get_conv_unary_pattern(
```
can you add some comments for this

leslie-fang-intel replied:
Thanks and added.

leslie-fang-intel committed:
Update on "Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap"


**Summary**
As per the discussion with Jerry in https://github.com/pytorch/pytorch/pull/114540#discussion_r1406945902, we plan to refactor `X86InductorQuantizer` annotations with `SubgraphMatcherWithNameNodeMap`. This plan will be implemented through several PRs, with this one being the first, focusing on refactoring single conv2d annotations in PTQ.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantizer/utils.py:
```diff
@@ -47,3 +60,57 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
+    conv_fn: Callable,
+    has_bn: bool = False,  # Usually need for QAT pattern
+    unary_fn: Optional[Callable[[Any], Any]] = None,
+):
```
can you add a return type as well

leslie-fang-intel replied:
Thanks for the comment and added.

jerryzh168 commented on torch/ao/quantization/quantizer/utils.py:
```diff
@@ -47,3 +60,57 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
+    return _conv_unary
+
+
+def _generate_pattern_matcher_helper(pattern, example_inputs):
```
nit: type annotations

leslie-fang-intel replied:
Thanks and added.

jerryzh168 commented on torch/ao/quantization/quantizer/utils.py:
```diff
@@ -47,3 +60,57 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
+
+
+@functools.lru_cache(None)
+def _generate_conv2d_pattern_matcher():
```
nit: type annotations

leslie-fang-intel replied:
Thanks and added.

leslie-fang-intel committed:
Update on "Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap"


**Summary**
As per the discussion with Jerry in https://github.com/pytorch/pytorch/pull/114540#discussion_r1406945902, we plan to refactor `X86InductorQuantizer` annotations with `SubgraphMatcherWithNameNodeMap`. This plan will be implemented through several PRs, with this one being the first, focusing on refactoring single conv2d annotations in PTQ.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap"


**Summary**
As per the discussion with Jerry in https://github.com/pytorch/pytorch/pull/114540#discussion_r1406945902, we plan to refactor `X86InductorQuantizer` annotations with `SubgraphMatcherWithNameNodeMap`. This plan will be implemented through several PRs, with this one being the first, focusing on refactoring single conv2d annotations in PTQ.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
